sled-agent/src/opte.rs

```diff
@@ -0,0 +1,334 @@
+    Opte(#[from] opte_ioctl::Error),
+
+    #[error(transparent)]
+    Dladm(#[from] dladm::Error),
+
+    #[error(transparent)]
+    Zone(#[from] zone::Error),
+
+    #[error(transparent)]
+    AddrObj(#[from] addrobj::Error),
+}
+
+/// Return a 24-bit virtual network identififier (VNI) from a UUID.
+///
+/// This currently takes the first 24 bits of the UUID.
+pub fn uuid_to_vni(id: &Uuid) -> Vni {
+    let f = id.as_fields().0;
+    Vni::new(f >> 8).unwrap()
+}
+
+#[derive(Debug, Clone)]
+pub struct OptePortAllocator {
+    value: Arc<AtomicU64>,
+}
+
+impl OptePortAllocator {
+    pub fn new() -> Self {
+        Self { value: Arc::new(AtomicU64::new(0)) }
+    }
+
+    fn next(&self) -> String {
+        format!("opte{}", self.next_id())
+    }
+
+    fn next_id(&self) -> u64 {
+        self.value.fetch_add(1, Ordering::SeqCst)
+    }
+
+    pub fn new_port(
+        &self,
+        ip: IpAddr,
+        mac: MacAddr6,
+        subnet: IpNetwork,
+        vpc_id: &Uuid,
+        underlay_ip: Ipv6Addr,
+    ) -> Result<OptePort, Error> {
+        // TODO-completess: Remove IPv4 restrictions once OPTE supports virtual
+        // IPv6 networks.
+        let private_ip = match ip {
+            IpAddr::V4(ip) => Ok(ip),
+            IpAddr::V6(_) => Err(opte_ioctl::Error::InvalidArgument(
+                String::from("IPv6 not yet supported"),
+            )),
+        }?;
+        let gateway = Gateway::from_subnet(&subnet);
+        let gateway_ip = match gateway.ip {
+            IpAddr::V4(ip) => Ok(ip),
+            IpAddr::V6(_) => Err(opte_ioctl::Error::InvalidArgument(
+                String::from("IPv6 not yet supported"),
+            )),
+        }?;
+        let boundary_services = BoundaryServices::default();
+        let vni = uuid_to_vni(vpc_id);
+
+        let name = self.next();
+        let hdl = OpteHdl::open(OpteHdl::DLD_CTL)?;
+        hdl.create_xde(
+            &name,
+            MacAddr::from(mac.into_array()),
+            private_ip,
+            MacAddr::from(gateway.mac.into_array()),
+            gateway_ip,
+            boundary_services.ip,
+            boundary_services.vni,
+            vni,
+            underlay_ip,
+            /* passthru = */ false,
+        )?;
+
+        // Insert this port into the sled's virtual-to-physical address mapping.
+        //
+        // The v2p mapping maps the overlay/virtual IP address to the 3-tuple
+        // `(overlay_mac, underlay_ip, VNI)`. That is, it provides "physical"
+        // information for the virtual network. This is a bit confusing, since
+        // it provides the guest MAC, but the idea is that this is a "physical
+        // address" as seen by the guest. The underlay IP is an address on the
+        // physical underlay network, and the VNI is used cordone traffic for a
+        // specific VPC, and so also used at the underlay network.
+        //
+        // Note that the v2p mapping is _not_ a mapping from overlay to underlay
```

Comment:
I'm not sure I agree with this comment. v2p is more than _just_ a mapping from overlay to underlay, but it does map an overlay address onto an underlay address.

Reply:
Fair enough. I wrote this comment after clarifying my initial misunderstanding of this mapping. I had thought it was a mapping from the virtual network to the physical network, i.e, something like a map `{virtual_ip, virtual_mac} -> {physical_ip, physical_map}`.  It is really a mapping `{virtual_ip} -> {virtual_mac, physical_ip, VNI}`. I'm not sure I'd really call that a virtual-to-physical mapping at all, but the point I was trying to make is that it's not a mapping from L2/L3 information in the underlay to L2/L3 information in the underlay as I originally thought.

I'll take another pass at the comment, thanks!